### ISSUE TITLE
Decrease memory usage for unpacking Kafka headers

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#20958]
A customer recently discovered that our [suggested pattern for decoding Kafka headers](https://materialize.com/docs/sql/create-source/kafka/#headers)...

```
SELECT
    id,
    seller,
    item,
    (
        SELECT convert_from((h).value, 'utf8') AS client_id
        FROM unnest(headers) AS h
        WHERE (h).key = 'client_id'
    )
FROM kafka_metadata;

 id | seller |        item        | client_id
----+--------+--------------------+-----------
  2 |   1592 | Custom Art         |        23
  7 |   1509 | Custom Art         |        42
  3 |   1411 | City Bar Crawl     |      "\N"
```

...has horrifically bad memory usage. @sthm tracked this down to this query plan:

```
explain SELECT
    (
        SELECT convert_from((h).value, 'utf8') AS client_id1
        FROM unnest(headers) AS h
        WHERE (h).key = 'h1'
    ),
        (
        SELECT convert_from((h).value, 'utf8') AS client_id2
        FROM unnest(headers) AS h
        WHERE (h).key = 'h2'
    ),
        (
        SELECT convert_from((h).value, 'utf8') AS client_id3
        FROM unnest(headers) AS h
        WHERE (h).key = 'h3'
    )
FROM test;
                             Optimized Plan
------------------------------------------------------------------------
 Explained Query:                                                      +
   Return                                                              +
     Project (#4, #7, #10)                                             +
       Join on=(eq(#0, #2, #5, #8) AND eq(#1, #3, #6, #9)) type=delta  +
         ArrangeBy keys=[[#0, #1]]                                     +
           Get materialize.public.test                                 +
         ArrangeBy keys=[[#0, #1]]                                     +
           Union                                                       +
             Get l3                                                    +
             Map (null)                                                +
               Union                                                   +
                 Negate                                                +
                   Distinct group_by=[#0, #1]                          +
                     Project (#0, #1)                                  +
                       Get l3                                          +
                 Get l0                                                +
         ArrangeBy keys=[[#0, #1]]                                     +
           Union                                                       +
             Get l5                                                    +
             Map (null)                                                +
               Union                                                   +
                 Negate                                                +
                   Distinct group_by=[#0, #1]                          +
                     Project (#0, #1)                                  +
                       Get l5                                          +
                 Get l0                                                +
         ArrangeBy keys=[[#0, #1]]                                     +
           Union                                                       +
             Get l7                                                    +
             Map (null)                                                +
               Union                                                   +
                 Negate                                                +
                   Distinct group_by=[#0, #1]                          +
                     Project (#0, #1)                                  +
                       Get l7                                          +
                 Get l0                                                +
   With                                                                +
     cte l7 =                                                          +
       Union                                                           +
         Project (#0, #1, #3)                                          +
           Get l6                                                      +
         Map (error("more than one record produced in subquery"))      +
           Project (#0, #1)                                            +
             Filter (#2 > 1)                                           +
               Reduce group_by=[#0, #1] aggregates=[count(*)] monotonic+
                 Project (#0, #1)                                      +
                   Get l6                                              +
     cte l6 =                                                          +
       Filter ("h3" = record_get[0](#2))                       +
         Get l1                                                        +
     cte l5 =                                                          +
       Union                                                           +
         Project (#0, #1, #3)                                          +
           Get l4                                                      +
         Map (error("more than one record produced in subquery"))      +
           Project (#0, #1)                                            +
             Filter (#2 > 1)                                           +
               Reduce group_by=[#0, #1] aggregates=[count(*)] monotonic+
                 Project (#0, #1)                                      +
                   Get l4                                              +
     cte l4 =                                                          +
       Filter ("h2" = record_get[0](#2))                       +
         Get l1                                                        +
     cte l3 =                                                          +
       Union                                                           +
         Project (#0, #1, #3)                                          +
           Get l2                                                      +
         Map (error("more than one record produced in subquery"))      +
           Project (#0, #1)                                            +
             Filter (#2 > 1)                                           +
               Reduce group_by=[#0, #1] aggregates=[count(*)] monotonic+
                 Project (#0, #1)                                      +
                   Get l2                                              +
     cte l2 =                                                          +
       Filter ("h1" = record_get[0](#2))                       +
         Get l1                                                        +
     cte l1 =                                                          +
       Map (convert_from(record_get[1](#2), "utf8"))                   +
         FlatMap unnest_list(#1)                                       +
           Get l0                                                      +
     cte l0 =                                                          +
       Distinct group_by=[#0, #1] monotonic                            +
         Get materialize.public.test                                   +

(1 row)
```

Slack conversation: https://materializeinc.slack.com/archives/C04DJT084KA/p1690561225164509

The problem is that the `unnest` call duplicates the entire row for each header value. With large rows and many header values, this is very expensive.

We introduced the `unnest` header decoding pattern in https://github.com/MaterializeInc/materialize/pull/20153. One option is to switch the docs back to the old style of header decoding by position:

```
staging > explain SELECT
       convert_from(headers[1].value, 'utf8') AS h1,
       convert_from(headers[2].value, 'utf8') AS h2,
       convert_from(headers[3].value, 'utf8') AS h3
FROM test;
                                                                                  Optimized Plan
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Explained Query:                                                                                                                                                                +
   Project (#2..=#4)                                                                                                                                                             +
     Map (convert_from(record_get[1](list_index(#1, 1)), "utf8"), convert_from(record_get[1](list_index(#1, 2)), "utf8"), convert_from(record_get[1](list_index(#1, 3)), "utf8"))+
       Get materialize.public.test                                                                                                                                               +

(1 row)
```

This is a bit scary, though, as there is no guarantee that Kafka headers are written in a consistent order across messages. That's why we changed to use `unnest` in the first place.

There are at least four alternatives I can think of:

1. Support `INCLUDE HEADERS MAP`, which produces a `headers` column as a `map[text => text]` rather than a `list[(key text, value text)]`. This would not support messages with multiple headers with the same key, and we'd need to define some semantics about duplicate keys (first wins or last wins).
2. Support `INCLUDE HEADERS ('h1' AS h1_alias, 'h2' AS h2_alias, ...)`, inspired by webhook sources, which provides individual header values as text under the provided names. We'd again need to define semantics for duplicate keys (first wins or last wins).
3. Support a `map_build` function that constructs a `map[text => T]` value out of a `list[(text, T)]` value. Then you could write `SELECT map_build(headers)['h1'], map_build(headers)['h2'] FROM kafka_src INCLUDE HEADERS`.
4. Teach the optimizer to better optimize the existing queries. I think in theory there's some projection pushdown magic that would make the existing query efficient.

(1) and (3) have the defect that they must store all the header values, even irrelevant ones, in the map. (4) is hard, and it's not like the current query is super ergonomic.

Of the options, (2) is my favorite. It is the most efficient _and_ makes for a very ergonomic query:

```
CREATE SOURCE kafka_src FROM KAFKA BROKER ... INCLUDE HEADERS ('h1' AS h1, 'h2' AS h2);
SELECT h1, h2, payload FROM kafka_src;
```

**Comment by frankmcsherry** [comment]
> The problem is that the unnest call duplicates the entire row for each header value. With large rows and many header values, this is very expensive.

Fwiw, unless I misunderstand I think the problem is really that decorrelation treats this as an arbitrary subquery, when in fact it is nicely scoped to individual records and could in principle be implemented by a `FlatMap`. The "problem" I think is both that 1. we don't recognize this (and in other similar cases) and 2. we don't have a scalar language rich enough to perform the evaluation. These both seem fixable, 1. through more careful lowering, and 2. through using the constant folding transform as an evaluator.

Don't let this get in the way of a more direct solution for the customer problem, but e.g. we had the same issue with @mjibson's Eve Skunkwoks, where each record had a list of things and we needed to apply `Distinct` to each list: there was no solution other than to flatten it to a large stateful `Distinct` and then reconstitute the results, when what one would have like to happen is a record-by-record `Map` or `FlatMap` that applies `list.distinct()`.

**Comment by ggevay** [comment]
> through using the constant folding transform as an evaluator.

Oh, you mean to have a `MirScalarExpr` variant whose evaluation would be `MirRelationExpr`'s constant folding? Interesting! So, the new `MirScalarExpr` variant would hold a `MirRelationExpr`, whose only input would be a placeholder relation, and this placeholder would be replaced by the evaluation every time by a `MirRelationExpr::Constant` that is the result of `unnest(headers)` on the current row of outer, right?

**Comment by frankmcsherry** [comment]
Right!

**Comment by ggevay** [comment]
Opened an issue for 4. based on the above discussion: https://github.com/MaterializeInc/materialize/issues/21787. But I think 2. is still better for this specific issue, because the syntax is nicer, and also it would still be more efficient CPU-wise: the approach proposed in https://github.com/MaterializeInc/materialize/issues/21787 would introduce as many FlatMaps as the number of columns that we are unpacking, and the `FoldConstants` calls would also be not super cheap.

**Comment by bkirwi** [comment]
Not like we need more options here, but...

Ran into postgres' `array_position` this morning, and it occurs to me that if we exposed the header keys and values as separate arrays (but with the same length) you could use normal postgres functions to deal with this... `array_position` to get the offset of the first matching header key and a normal lookup in the header values.

**Comment by sthm** [comment]
Closed by #22806 and https://github.com/MaterializeInc/materialize/pull/23419.

**Comment by benesch** [comment]
Going to close this as completed, since the issue didn't mean to indicate that we do _all_ the options—just one! To summarize:

> 1. Support `INCLUDE HEADERS MAP`, which produces a `headers` column as a `map[text => text]` rather than a `list[(key text, value text)]`. This would not support messages with multiple headers with the same key, and we'd need to define some semantics about duplicate keys (first wins or last wins).

Since we did (2), I think we will never want to do this, because that would result in three separate ways of getting headers. That's too many.

> 2. Support `INCLUDE HEADERS ('h1' AS h1_alias, 'h2' AS h2_alias, ...)`, inspired by webhook sources, which provides individual header values as text under the provided names. We'd again need to define semantics for duplicate keys (first wins or last wins).

This is done! Thanks, @sthm! 🎉 

> 3. Support a `map_build` function that constructs a `map[text => T]` value out of a `list[(text, T)]` value. Then you could write `SELECT map_build(headers)['h1'], map_build(headers)['h2'] FROM kafka_src INCLUDE HEADERS`.

I think this function could plausibly be useful, but no pressing need for it, so I'm not going to bother filing a follow up until there's further demand.

> 4. Teach the optimizer to better optimize the existing queries. I think in theory there's some projection pushdown magic that would make the existing query efficient.

This is definitely worthwhile, as this pattern of "operate on a row efficiently" comes up in other contexts. Fortunately we've already got https://github.com/MaterializeInc/materialize/issues/21787 to track.
